### ISSUE TITLE
Unified HaloMassFunction with numpy/torch/jax backends

### ISSUE THREAD

**Issue by sambit-giri** [cosmic-reionization/BEoRN#39]
## Motivation

BEoRN currently has HMF logic in three places:

- `beorn/mass_function/models.py` — `PressSchechter`, `ShethTormen` (BEoRN-native, numpy)
- `beorn/mass_function/toolscosmo_hmf.py` — wraps `toolscosmo.massfct.fnu`
- `beorn/lpt/chmf.py` — `CHMF` re-implements PS f(ν) inline for conditional sampling

This issue plans a single `HaloMassFunction` class that consolidates all three, adds window-function flexibility, and supports autodiff backends.

---

## Proposed API

```python
from beorn.mass_function import HaloMassFunction

# Named model (default)
hmf = HaloMassFunction(param, model='ST', window='tophat', backend='numpy')

# Generic parametric — any (A, p, q) in the unified f(ν) family
hmf = HaloMassFunction(param, model='custom', A=0.3, p=0.3, q=0.8)

# Differentiable pipeline
hmf = HaloMassFunction(param, model='ST', backend='jax')

# Evaluate
dn_dlnM = hmf.dndlnm(M, z)   # (Mpc/h)^{-3}
sigma2  = hmf.sigma2(M, z)
```

Convenience runners for discoverability:

```python
from beorn.mass_function import run_PressSchechter, run_ShethTormen, run_Jenkins

dn = run_ShethTormen(param, M, z)
```

---

### Parametric f(ν) family — same formula, different (A, p, q)

Many models share the unified first-crossing distribution

```
f(ν) = A · √(2qν/π) · (1 + (qν)^{−p}) · exp(−qν/2)    where ν = δ_c² / σ²
```

| Model | A | p | q | Window |
|---|---|---|---|---|
| `'PS'`  Press-Schechter (1974) | 0.5 | 0 | 1 | tophat |
| `'ST'`  Sheth-Tormen (2002) | ≈ 0.3222 | 0.3 | 0.707 | tophat |
| `'EC'`  Ellipsoidal collapse | ≈ 0.3222 | 0.3 | 1 | tophat |
| `'custom'` / `None` | user | user | user | any |

`A` is self-consistently normalised as `1 / (1 + 2^{−p} Γ(0.5−p) / √π)` unless overridden.

### Simulation-calibrated models — different functional forms

These do **not** fit the parametric f(ν) above and require separate implementations:

| Model | Form | Reference |
|---|---|---|
| `'Jenkins'` | `A exp(−\|ln σ⁻¹ + B\|^ε)` | Jenkins et al. (2001) |
| `'Watson'`  | Jenkins-like with redshift-dependent coefficients | Watson et al. (2013) |
| `'Tinker08'`| `A((σ/b)^{−a} + 1) exp(−c/σ²)` | Tinker et al. (2008) |

---

## Window functions for σ²(M)

The variance σ²(M) = (1/2π²) ∫ k³ P(k) W²(kR) d ln k depends on the window W(kR):

| `window=` | W(kR) | Notes |
|---|---|---|
| `'tophat'` | 3(sin x − x cos x)/x³ | Standard; W(0)=1, sharp in real space |
| `'sharp-k'` | Θ(1 − kR) | Step in k; used by Jenkins, some EPS variants |
| `'smooth-k'` | 1/(1 + (kR)^β) | Smooth cutoff; allows analytical d ln σ/d ln M |

The choice of window affects both σ²(M) and the M(R) relation, so model and window should be coupled where physically motivated (e.g. Jenkins uses sharp-k).

---

## Backend design

The σ²(M) integral and f(ν) formula are algebraically simple — both are
differentiable with respect to cosmological parameters (Ωm, σ₈, ns, …).

```
backend='numpy'  →  current behaviour, no extra deps
backend='torch'  →  torch.trapezoid, torch.exp, etc.; gradients via autograd
backend='jax'    →  jnp.trapezoid, jax.grad; integrates with existing JaxBackend
```

Key insight: BEoRN already has `JaxBackend` for the LPT density field.
A JAX HMF would complete the differentiable chain

```
θ_cosmo  →  P(k, θ)  →  σ²(M, θ)  →  dn/d ln M  →  n_halo(z)  →  τ_e / 21-cm PS
```

enabling gradient-based inference (SBI, HMC, variational) over EoR observables.

---

## File structure

```
src/beorn/mass_function/
    __init__.py
    base.py              ← MassFunction ABC (sigma²; backend-agnostic interface)
    sigma.py             ← sigma²(M, z) for tophat / sharp-k / smooth-k × numpy/torch/jax
    parametric.py        ← ParametricHMF (the f(ν) family); replaces current models.py
    calibrated.py        ← Jenkins, Watson, Tinker — non-parametric fitted models
    runners.py           ← run_PressSchechter(), run_ShethTormen(), run_Jenkins(), …
    toolscosmo_hmf.py    ← kept as thin shim; delegates f(ν) to toolscosmo
```

---

## Migration

| Current location | Target |
|---|---|
| `beorn/lpt/chmf.py` — `CHMF._precompute_sigma` | `sigma.py` |
| `beorn/lpt/chmf.py` — `hmf_ps` | `parametric.PressSchechter` |
| `beorn/mass_function/models.py` | merged into `parametric.py` |
| `beorn/plotting/halo_mass_function.py` — `hmf` pkg | optional; `calibrated.Tinker08` |

`CHMF` in `chmf.py` stays as the conditional sampler but imports its f(ν) from `parametric.ParametricHMF` instead of re-implementing it inline.

---

## Out of scope for this issue

- Conditional HMF (EPS) sampling — already in `beorn/lpt/chmf.py`
- Halo bias
- Merger trees / mass accretion

**Comment by sambit-giri** [comment]
## Implemented — commit 53a635d on `v2-review`

All items from the proposal have been delivered:

### `beorn.mass_function.HaloMassFunction`
- Unified entry point for PS, ST, and ellipsoidal collapse
- Pre-computes σ²(M) once on construction via `_sigma2_interp` (log-log interpolation)
- `dndlnm(M, z, ...)` for evaluation; `run_press_schechter`, `run_sheth_tormen`, `run_ellipsoidal` convenience runners
- Window functions: `tophat`, `sharp_k`, `smooth_k` (β=4 default)
- Backends: `numpy` (default), `jax`, `torch`

### Autodiff
- JAX: `jax.grad` / `jax.jacobian` through δ_c, p, q (verified vs finite differences, error < 4 × 10⁻⁹)
- PyTorch: `requires_grad=True` on δ_c, p, q — JAX and torch agree to < 2 × 10⁻¹⁶
- Both backends use the same pre-computed numpy σ²(M) grid; only f(ν) runs differentiably

### `TabulatedPowerSpectrum` (bonus, motivated by HMF validation)
- Accepts any external P(k) table (colossus, CAMB, etc.) via `beorn.lpt.TabulatedPowerSpectrum(param, k, Pk)`
- Optional σ₈ renormalisation and BEoRN growth-factor redshift evolution
- Plugs directly into `HaloMassFunction(param, ps_method='tabulated', k=..., Pk=...)`

### E&H wiggle fix
- Corrected Silk damping formula, Γ_eff units (Ωm h, not Ωm h²), and h/Mpc → Mpc⁻¹ unit conversion for the wiggle transfer function

### Validation notebook
`examples/ps_hmf_models.ipynb` covers:
- PS and ST HMF vs colossus with matched P(k) — ratio within ~1% across 10⁸–10¹³ M☉ once identical P(k) is used
- JAX gradient section with finite-difference verification
- PyTorch autograd section matching JAX gradients to machine precision

### File structure (matches proposal)
```
src/beorn/mass_function/
    __init__.py     ← exports HaloMassFunction, PressSchechter, ShethTormen, window classes
    base.py         ← MassFunction ABC (sigma² + window-agnostic interface)
    models.py       ← HaloMassFunction, PressSchechter, ShethTormen, ellipsoidal
    window.py       ← TopHatWindow, SharpKWindow, SmoothKWindow
tests/test_mass_function.py
```

The remaining items (Jenkins/Watson/Tinker calibrated models, halo bias, merger trees) were listed as out-of-scope in the original issue and remain deferred.
